### ISSUE TITLE
Remote attestation fails - Error: SGX_QL_ERROR_INVALID_PRIVILEGE

### ISSUE THREAD

**Issue by rockhouse** [edgelesssys/ego#159]
### Issue description
Followed the tutorial to get the TLS attestation sample working. 
Local machine deployment not cloud. 
Have PCCS running, docker logs output:
```
Generating RSA private key, 2048 bit long modulus (2 primes)
..................................................................+++++
..............................+++++
e is 65537 (0x010001)
Signature ok
subject=CN = localhost
Getting Private key
Wed, 03 Aug 2022 12:49:45 GMT morgan deprecated default format: use combined format at node_modules/esm/esm.js:1:278827
2022-08-03 12:49:46.871 [info]: DB Migration (Ver.0 -> 1) -- Start
2022-08-03 12:49:46.879 [info]: DB Migration -- Done.
2022-08-03 12:49:46.935 [info]: DB Migration (Ver.1 -> 2) -- Start
2022-08-03 12:49:46.948 [info]: DB Migration -- Done.
2022-08-03 12:49:47.043 [info]: HTTPS Server is running on: https://localhost:8081
2022-08-03 13:04:36.004 [info]: Client Request-ID : 07a4756340594e0ab7c84bca10d56f79
2022-08-03 13:04:36.006 [info]: 172.17.0.1 - - [03/Aug/2022:13:04:36 +0000] "GET /sgx/certification/v3/rootcacrl HTTP/1.1" 200 586 "-" "curl/7.58.0"
```
As you can see also tried to curl the root ca which worked. 

Error I am seeing is:
```
EGo v1.0.0 (f1255317ec583ed72947f65d83881a0e46ad1ed8)
[erthost] loading enclave ...
[erthost] entering enclave ...
[ego] starting application ...
[load_qe ../qe_logic.cpp:642] Error, call sgx_create_enclave QE fail [load_qe], SGXError:4004.
ERROR: quote3_error_t=SGX_QL_ERROR_INVALID_PRIVILEGE
 (oe_result_t=OE_PLATFORM_ERROR) [openenclave-src/host/sgx/sgxquote.c:oe_sgx_qe_get_target_info:706]
ERROR: SGX Plugin _get_report(): failed to get ecdsa report. OE_PLATFORM_ERROR (oe_result_t=OE_PLATFORM_ERROR) [openenclave-src/enclave/sgx/attester.c:_get_report:324]
panic: OE_PLATFORM_ERROR

goroutine 1 [running]:
main.main()
	/root/golang/ego/samples/attested_tls/server.go:14 +0x192
```

I have seen the earlier issue https://github.com/edgelesssys/ego/issues/158 but as the error is a different one I dont think it is about a BIOS update.

**Comment by thomasten** [comment]
Hi,
Can you try running it with sudo?

**Comment by rockhouse** [comment]
I am calling the `ego run server` as user root

**Comment by thomasten** [comment]
I need to do some research and then will get back to you.

**Comment by EErikas** [comment]
I have tried the same example and gotten the same exact output. @rockhouse could you check if you face similar issues with `remote_attestation`  sample as described in #158?  Bios update did not solve the issue for me.

**Comment by rockhouse** [comment]
I have re-read your [documentation](https://docs.edgeless.systems/ego/#/getting-started/troubleshoot) and found this passage: 
```
$ cpuid | grep SGX
      SGX: Software Guard Extensions supported = true
      SGX_LC: SGX launch config supported      = true
```

for our machine SGX_LC is actually **false** only the first is set to true. It seems not to be a BIOS option to switch on. So question is, is this required for the remote attestation? If yes and it is not a BIOS option how can one enable it? 

[Here](https://techlibrary.hpe.com/docs/iss/proliant-gen10-uefi/GUID-610261F0-3E25-4E90-8906-297D5E77E521.html) are the SGX BIOS options that are available for our motherboard. There is no option regarding "launch config" while SGX is clearly supported. 

@EErikas I tried [remote_attestation](https://github.com/edgelesssys/ego/tree/master/samples/remote_attestation) and [attested_tls](https://github.com/edgelesssys/ego/tree/master/samples/attested_tls) both fail while PCCS output looks fine. 

Any help is appreciated.

**Comment by rockhouse** [comment]
The initial issue seems to be solved, we finally got our hands on a machine where SGX_LC: SGX launch config is supported. 

We then encountered a new error:
```
[get_platform_quote_cert_data ../qe_logic.cpp:347] Error returned from the p_sgx_get_quote_config API. 0xe019
ERROR: quote3_error_t=SGX_QL_NETWORK_ERROR
 (oe_result_t=OE_PLATFORM_ERROR) [openenclave-src/host/sgx/sgxquote.c:oe_sgx_qe_get_target_info:706]
ERROR: SGX Plugin _get_report(): failed to get ecdsa report. OE_PLATFORM_ERROR (oe_result_t=OE_PLATFORM_ERROR) [openenclave-src/enclave/sgx/attester.c:_get_report:324]
panic: OE_PLATFORM_ERROR
```

But that could be solved by setting  "use_secure_cert" to **false** in /etc/sgx_default_qcnl.conf as described [here](https://github.com/openenclave/openenclave/blob/master/docs/GettingStartedDocs/Contributors/NonAccMachineSGXLinuxGettingStarted.md#33-configure-the-qpl)
Would be interesting to know how to get a proper certificate for the pccs service. I am actually also not sure if this is a proper solution to the mentioned error. 

Even after managing this issue we unfortunately encountered yet another one which is when using the client:

This is the output of the enclave:
```
EGo v1.0.0 (xxx)
[erthost] loading enclave ...
[erthost] entering enclave ...
[ego] starting application ...
listening ...
2022/08/12 14:24:52 http: TLS handshake error from [::1]:53602: remote error: tls: bad certificate
```
And this from the client:
```
EGo v1.0.0 (xxxx)
2022-08-12T09:24:52-0500.530429Z [(H)ERROR] tid(0x7f18df7fe700) | Failed to read CRL. OE_CRYPTO_ERROR (oe_result_t=OE_CRYPTO_ERROR) [/ertbuild/3rdparty/openenclave/openenclave-src/common/sgx/collateral.c:oe_validate_revocation_list:385]
2022-08-12T09:24:52-0500.530490Z [(H)ERROR] tid(0x7f18df7fe700) | :OE_INVALID_PARAMETER [/ertbuild/3rdparty/openenclave/openenclave-src/common/crypto/openssl/crl.c:oe_crl_free:155]
2022-08-12T09:24:52-0500.530502Z [(H)ERROR] tid(0x7f18df7fe700) | :OE_INVALID_PARAMETER [/ertbuild/3rdparty/openenclave/openenclave-src/common/crypto/openssl/crl.c:oe_crl_free:155]
2022-08-12T09:24:52-0500.530581Z [(H)ERROR] tid(0x7f18df7fe700) | :OE_INVALID_PARAMETER [/ertbuild/3rdparty/openenclave/openenclave-src/common/crypto/openssl/cert.c:oe_cert_free:578]
2022-08-12T09:24:52-0500.530598Z [(H)ERROR] tid(0x7f18df7fe700) | Failed to validate revocation info. OE_CRYPTO_ERROR (oe_result_t=OE_CRYPTO_ERROR) [/ertbuild/3rdparty/openenclave/openenclave-src/common/sgx/quote.c:oe_get_sgx_quote_validity:776]
2022-08-12T09:24:52-0500.530673Z [(H)ERROR] tid(0x7f18df7fe700) | Failed to validate quote. OE_CRYPTO_ERROR (oe_result_t=OE_CRYPTO_ERROR) [/ertbuild/3rdparty/openenclave/openenclave-src/common/sgx/quote.c:oe_verify_quote_with_sgx_endorsements:631]
2022-08-12T09:24:52-0500.530688Z [(H)ERROR] tid(0x7f18df7fe700) | :OE_CRYPTO_ERROR [/ertbuild/3rdparty/openenclave/openenclave-src/common/sgx/verifier.c:oe_sgx_verify_evidence:900]
2022-08-12T09:24:52-0500.530699Z [(H)ERROR] tid(0x7f18df7fe700) | :OE_CRYPTO_ERROR [/ertbuild/3rdparty/openenclave/openenclave-src/common/attest_plugin.c:oe_verify_evidence:438]
panic: Get "https://localhost:8080/secret?s=mySecret": OE_CRYPTO_ERROR

goroutine 1 [running]:
main.httpGet(0xc00014a000, {0xc00001e2d0, 0x28})
	/root/ego/samples/attested_tls/ra_client/client.go:65 +0x205
main.main()
	/root/ego/samples/attested_tls/ra_client/client.go:39 +0x1b6
```

This is the output of the pccs:
```
2022-08-12 14:27:00.878 [info]: Client Request-ID : df4ad40f2e984570bbc51b3707269a0e
2022-08-12 14:27:00.881 [info]: 172.17.0.1 - - [12/Aug/2022:14:27:00 +0000] "GET /sgx/certification/v3/pckcrl?ca=processor HTTP/1.1" 200 606 "-" "-"

2022-08-12 14:27:00.899 [info]: Client Request-ID : 469cfea434eb42678a6fcf73c9b7458c
2022-08-12 14:27:00.901 [info]: 172.17.0.1 - - [12/Aug/2022:14:27:00 +0000] "GET /sgx/certification/v3/tcb?fmspc=00906EA10000 HTTP/1.1" 200 7725 "-" "-"

2022-08-12 14:27:00.919 [info]: Client Request-ID : c5e7b87a4af0447088c2f90a18d1dfcd
2022-08-12 14:27:00.921 [info]: 172.17.0.1 - - [12/Aug/2022:14:27:00 +0000] "GET /sgx/certification/v3/qe/identity HTTP/1.1" 200 933 "-" "-"

2022-08-12 14:27:00.939 [info]: Client Request-ID : b67955278c3a4d02ad21ae94d1057830
2022-08-12 14:27:00.944 [info]: 172.17.0.1 - - [12/Aug/2022:14:27:00 +0000] "GET /sgx/certification/v3/rootcacrl HTTP/1.1" 200 586 "-" "-"
```

Not sure how to proceed now.

**Comment by thomasten** [comment]
> The initial issue seems to be solved, we finally got our hands on a machine where SGX_LC: SGX launch config is supported.

Great!

> Would be interesting to know how to get a proper certificate for the pccs service.

This is a standard PKI issue with no SGX specifics. The PCCS that we provide as a docker image generates a self-signed certificate for simplicity. You could replace this by a certificate signed by your organization's CA.

> I am actually also not sure if this is a proper solution to the mentioned error.

As long as you only use the PCCS for pulling the required certificates and collaterals (via the quote provider library), this is fine. The connection to the PCCS doesn't need to be secure because the received data is verified anyway.
In addition, the PCCS has an API to configure it. This requires to send a password to it and this shouldn't be done without properly configured TLS. By default, our PCCS container generates a random password to effectively prevent any access.

> we unfortunately encountered yet another one which is when using the client

There's currently a bug in the underlying Open Enclave that breaks PCCS API v3.0. We'll fix this with the next release of EGo. As a workaround, you can set `pccs_api_version` to `3.1` in sgx_default_qcnl.conf.

**Comment by rockhouse** [comment]
Thanks @thomasten for your support and help, it works now ;)

**Comment by ylht** [comment]
> > we unfortunately encountered yet another one which is when using the client
> 
> There's currently a bug in the underlying Open Enclave that breaks PCCS API v3.0. We'll fix this with the next release of EGo. As a workaround, you can set `pccs_api_version` to `3.1` in sgx_default_qcnl.conf.

This solution successfully solved my problem, do you have time to add this solution in `README` of remote attestion? It could effectively help other people who are ready to try to use `ego`.

**Comment by thomasten** [comment]
PCCS API v3.0 is fixed in EGo 1.0.1

**Comment by ylht** [comment]
> PCCS API v3.0 is fixed in EGo 1.0.1

Thanks
